Improve handling for sized string return values

cgrenz opened the pull request:
This PR fixes the second problem mentioned in issue #10:
functions that return strings and have an unsigned*/size_t*
argument in which the string length is written.

All such functions have the length out-argument as the last argument,
so I limited the detection to only check the last argument.

aleclearmind commented on llvmcpy/llvm.py:
```diff
@@ -257,6 +258,28 @@ def call():
+          and return_type.item.cname == "char"
+          and last_arg and last_arg.kind == "pointer"
+          and last_arg.item.kind == "primitive"
+          and last_arg.item.cname in ("unsigned int", "unsigned long")):
```
Maybe we could have a global list of "integer" types such as `signed`, `unsigned`, `long`, `int`, `long long`, `long int`, `unsigned int`, `signed int`...

cgrenz replied:
All arguments used for lengths are `unsigned` or `size_t` (which always maps to an unsigned integer type), so two separate lists for unsigned and signed integer types might be better.

aleclearmind replied:
ACK

aleclearmind commented on llvmcpy/llvm.py:
```diff
@@ -257,6 +258,28 @@ def call():
+        # This often represents returning a string (that may contain
+        # NUL characters) with a length
+        function_arguments[-1] = None
+        arguments[-1] = "lengthbuf"
```
Can we name the argument `buffer_length`?

cgrenz replied:
It's a buffer to hold the length of the returned string, so `length_buffer` might be more clear.

aleclearmind replied:
ACK

aleclearmind commented on llvmcpy/llvm.py:
```diff
@@ -257,6 +258,28 @@ def call():
+        result += """        lengthbuf = ffi.new("{}[1]")
+        ptr = {}
+        length = lengthbuf[0]
+        return ffi.unpack(ptr, length)""".format(last_arg.item.cname, call())
```
Any specific reason for using `ffi.unpack` instead of `ffi.string` as we do elsewhere?

cgrenz replied:
Yes. `ffi.string` uses C-string semantics, so a NUL character is interpreted as the end of the string. `ffi.unpack` converts the whole string up to pointer+length into a bytestring

aleclearmind replied:
Right, we should check elsewhere if there's some erroneous usage of `ffi.string`, but I suppose they're fine.

aleclearmind commented:
Thanks for the patch and sorry for the delay! Can you please address the comments and maybe add a quick test?

aleclearmind commented:
* `ffi.unpack` and `ffi.string` return a byte array in Python 3, can you make a commit to add `.decode("utf-8")` to all of their occurrences? We already do this elsewhere. See the failing tests.
* The test case fails with LLVM 3.4 and 3.5, it seems they don't have `LLVMGetAsString`, if you can find another function to use it would be nice.
  Otherwise you can use `llvm.version` to check if the version is greater than 3.5, if you do so, please add `.decode("utf-8")` to the version variable too. You could use [LooseVersion](http://stackoverflow.com/questions/11887762/compare-version-strings-in-python/21065570#21065570).
  Last option: skip the test if `value.get_as_string is None`.
* Can you squash the other commits in a single one?

Thanks.

cgrenz commented:
* decoding to `str` using utf-8 might not always be the right choice as returned strings might contain binary data. I added an optional argument `encoding` to allow setting another encoding or getting a byte string.
* I switched the test case to use LLVMMDString()/LLVMGetMDString(). These are supported since LLVM 3.0.
* Can I get github to squash this pull request or what do I have to do? I don't usually use github.

aleclearmind commented:
1. Uhm, you're right about strings. Your choice seems reasonable to me, let's go for it.
2. Cool.
3. You can use `git rebase -i` to rewrite your branch and [squash commits together](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html), then just `git push -f` and we should be fine. I'd like to have two commits: one to fix past stuff (i.e., 4fed92c34c5d3f98de61a4060e7bbbae033c5605) and one for all the rest.

cgrenz commented:
Ok, so force pushing works with github. Good to know :-)

I'm a bit confused. What do you want to have split in two commits?
I didn't change past stuff in the commits in this pull request, I only added code for the sized string returns
(see `git diff 679d2c1 e876369`).

aleclearmind commented:
I see 7 commits right now on the cgrenz:string-length-args, I'd like only two of them: one to fix past stuff and one for all the rest. You can do that by rewriting the cgrenz:string-length-args branch and then `-f` pushing it to GH. I can do it for you otherwise!

aleclearmind commented:
Merged in e27b4547559d9cbd89a905f422cf66c4a98ca7c1 with minor changes. Thanks a lot!
